scene/gui/virtual_joystick.cpp

```diff
@@ -0,0 +1,381 @@
+
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "joystick_texture", PROPERTY_HINT_RESOURCE_TYPE, "Texture2D"), "set_joystick_texture", "get_joystick_texture");
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "tip_texture", PROPERTY_HINT_RESOURCE_TYPE, "Texture2D"), "set_tip_texture", "get_tip_texture");
+
+	BIND_ENUM_CONSTANT(JOYSTICK_FIXED);
+	BIND_ENUM_CONSTANT(JOYSTICK_DYNAMIC);
+	BIND_ENUM_CONSTANT(JOYSTICK_FOLLOWING);
+	BIND_ENUM_CONSTANT(VISIBILITY_ALWAYS);
+	BIND_ENUM_CONSTANT(VISIBILITY_TOUCHSCREEN_ONLY);
+	BIND_ENUM_CONSTANT(VISIBILITY_WHEN_TOUCHED);
+
+	BIND_THEME_ITEM(Theme::DATA_TYPE_COLOR, VirtualJoystick, base_normal_color);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_COLOR, VirtualJoystick, tip_normal_color);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_COLOR, VirtualJoystick, base_pressed_color);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_COLOR, VirtualJoystick, tip_pressed_color);
+}
+
+void VirtualJoystick::set_joystick_size(float p_size) {
+	joystick_size = p_size;
+	clampzone_size = joystick_size;
+	_reset();
+}
+
+float VirtualJoystick::get_joystick_size() const {
+	return joystick_size;
+}
+
+void VirtualJoystick::set_deadzone_size(float p_size) {
+	deadzone_size = p_size;
+}
+
+float VirtualJoystick::get_deadzone_size() const {
+	return deadzone_size;
+}
+
+void VirtualJoystick::set_clampzone_size(float p_size) {
+	clampzone_size = p_size;
+}
+
+float VirtualJoystick::get_clampzone_size() const {
+	return clampzone_size;
+}
+
+void VirtualJoystick::set_initial_offset_ratio(const Vector2 &p_ratio) {
+	initial_offset_ratio = p_ratio;
+	_reset();
+}
+
+Vector2 VirtualJoystick::get_initial_offset_ratio() const {
+	return initial_offset_ratio;
+}
+
+void VirtualJoystick::set_joystick_mode(JoystickMode p_mode) {
+	joystick_mode = p_mode;
+}
+
+VirtualJoystick::JoystickMode VirtualJoystick::get_joystick_mode() const {
+	return joystick_mode;
+}
+
+void VirtualJoystick::set_action_left(const String &p_action) {
+	action_left = p_action;
+}
+
+String VirtualJoystick::get_action_left() const {
+	return action_left;
+}
+
+void VirtualJoystick::set_action_right(const String &p_action) {
+	action_right = p_action;
+}
+
+String VirtualJoystick::get_action_right() const {
+	return action_right;
+}
+
+void VirtualJoystick::set_action_up(const String &p_action) {
+	action_up = p_action;
+}
+
+String VirtualJoystick::get_action_up() const {
+	return action_up;
+}
+
+void VirtualJoystick::set_action_down(const String &p_action) {
+	action_down = p_action;
+}
+
+String VirtualJoystick::get_action_down() const {
+	return action_down;
+}
+
+void VirtualJoystick::set_visibility_mode(VisibilityMode p_mode) {
+	visibility = p_mode;
+}
+
+VirtualJoystick::VisibilityMode VirtualJoystick::get_visibility_mode() const {
+	return visibility;
+}
+
+void VirtualJoystick::set_joystick_texture(const Ref<Texture2D> &p_texture) {
+	joystick_texture = p_texture;
```

Comment:
Only update if texture changed
```suggestion
	if (joystick_texture == p_texture) {
		return;
	}
	joystick_texture = p_texture;
```
Same below.